File: CircleFigure.lua
-- [[ REMOVE THIS ]]--
function AddLightning (codeName, checkVisibility, x1, y1, x2, y2)
    return {}
end
function DestroyLightning (whichBolt)
end
function MoveLightning (whichBolt, checkVisibility, x1, y1, x2, y2)
end
-- [[ END REMOVE THIS ]]--


local function normalize (angle)
    return (angle % (2 * math.pi) + 2 * math.pi) % (2 * math.pi);
end

CircleFigureEdge = {}

---@param xa number
---@param ya number
---@param xb number
---@param yb number
---@param codeName string
function CircleFigureEdge:new(xa, ya, xb, yb, codeName)
    local t = {
        lightning = AddLightning(codeName, false, xa, ya, xb, yb)
    }
    setmetatable(t, self)
    self.__index = self
    self:position(xa, ya, xb, yb)
    return t
end

---@param xa number
---@param ya number
---@param xb number
---@param yb number
function CircleFigureEdge:position(xa, ya, xb, yb)
    self.xa, self.ya, self.xb, self.yb = xa, ya, xb, yb
    MoveLightning(self.lightning, xa, ya, xb, yb)
end

function CircleFigureEdge:destroy()
    DestroyLightning(self.lightning)
end

---@class CircleFigure
CircleFigure = {}

---@param x number
---@param y number
---@param radius number
---@param angle number
---@param count number
function CircleFigure:new(x, y, radius, angle, count, codeName)
    local t = {
        data = {},
        edges = {},
        codeName = codeName,
    };
    setmetatable(t, {
        __index = function(self, key)
            if key == 'x' or key == 'y' or key == 'radius' or key == 'angle' then
                return self.data[key];
            end
            return rawget(CircleFigure, key);
        end,
        __newindex = function(self, key, value)
            if key == 'x' or key == 'y' then
                self.data[key] = value;
                return self:update();
            end
            if key == 'radius' then
                self.data[key] = math.max(0, value)
                return self:update();
            end
            if key == 'angle' then
                self.data[key] = normalize(value)
                return self:update();
            end
            if key == 'count' then
                self.data[key] = math.max(3, value)
                return self:update();
            end
            rawset(self, key, value)
        end
    });
    t.x = x;
    t.y = y;
    t.radius = radius;
    t.angle = angle;
    t.count = count;
    return t;
end

function CircleFigure:update()
    local d = self.data;
    if d.x == nil or d.y == nil or d.radius == nil or d.angle == nil or d.count == nil then
        return
    end

    while #self.edges > 0 and #self.edges > d.count do
        table.remove(self.edges):destroy()
    end

    local ad = 2 * math.pi / d.count
    local a = d.angle
    local xa = d.x + d.radius + math.cos(a)
    local ya = d.y + d.radius * math.sin(a)
    local xb, yb

    for i = 1, d.count, 1 do
        a = a + ad;
        xb = d.x + d.radius * math.cos(a);
        yb = d.y + d.radius * math.sin(a);
        if #self.edges < i then
            table.insert(self.edges, CircleFigureEdge:new(xa, ya, xb, yb, self.codeName))
        else
            self.edges[i]:position(xa, ya, xb, yb)
        end
        xa, ya = xb, yb
    end
end

function CircleFigure:destroy()
    while #self.data.edges > 0 do
        table.remove(self.data.edges):destroy()
    end
end

local c = CircleFigure:new(10, 20, 50, 11, 1);
c.x = 10;
c.x = c.x + 10;
c.count = 4;
c.count = 2;

File: CircleFigure.mjs
const normalize = angle => (angle % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI);

class CircleFigureEdge {
    /**
     * @param {number} xa
     * @param {number} ya
     * @param {number} xb
     * @param {number} yb
     */
    constructor(xa, ya, xb, yb) {
        this.position(xa, ya, xb, yb);
    }

    /**
     * @param {number} xa
     * @param {number} ya
     * @param {number} xb
     * @param {number} yb
     */
    position(xa, ya, xb, yb) {
        this.xa = xa;
        this.ya = ya;
        this.xb = xb;
        this.yb = yb;
    }

    destroy() {
        console.log('☢️ Edge destroed!');
        return this;
    }
}

export class CircleFigure {

    /**
     * @param {number} x
     * @param {number} y
     * @param {number} count
     * @param {number} radius
     * @param {number} angle
     */
    constructor(x, y, radius, angle, count) {
        this.x = x;
        this.y = y;
        this.radius = radius;
        this.angle = angle;
        this.count = count;
        this.update();
    }

    /** @type {number} */ #x;
    /** @param {number} x */set x(x) {
        if (this.#x === x) return;
        this.#x = x;
        this.update();
    }

    /** @return {number} */get x() {
        return this.#x
    }

    /** @type {number} */ #y;
    /** @param {number} y */set y(y) {
        if (this.#y === y) return;
        this.#y = y;
        this.update();
    }

    /** @return {number} */get y() {
        return this.#y
    }

    /** @type {number} */ #radius;
    /** @param {number} radius */set radius(radius) {
        if (this.#radius === radius) return;
        this.#radius = Math.max(0, radius);
        this.update();
    }

    /** @return {number} */get radius() {
        return this.#radius
    }

    /** @type {number} */ #angle;
    /** @param {number} angle */set angle(angle) {
        if (this.#angle === angle) return;
        this.#angle = normalize(angle);
        this.update();
    }

    /** @return {number} */get angle() {
        return this.#angle
    }

    /** @type {number} */ #count;
    /** @param {number} count */set count(count) {
        if (this.#count === count) return;
        this.#count = Math.max(3, count);
        this.update()
    }

    /** @return {number} */get count() {
        return this.#count
    }

    /** @type {CircleFigureEdge[]} */edges = [];

    update() {
        if (this.#x === undefined || this.#y === undefined || this.#angle === undefined || this.#radius === undefined || this.#count === undefined) return;

        while (this.edges.length > 0 && this.edges.length > this.#count) {
            this.edges.pop().destroy();
        }
        const ad = 2 * Math.PI / this.count;
        let a = this.#angle;
        let xa = this.#x + this.#radius * Math.cos(a);
        let ya = this.#y + this.#radius * Math.sin(a);
        let xb, yb;
        for (let i = 0; i < this.count; i++) {
            a += ad;
            xb = this.#x + this.#radius * Math.cos(a);
            yb = this.#y + this.#radius * Math.sin(a);
            if (this.edges.length < i + 1) this.edges.push(new CircleFigureEdge(xa, ya, xb, yb));
            else this.edges[i].position(xa, ya, xb, yb);
            xa = xb;
            ya = yb;
        }
    }
}